lib/decompress.ps1

```diff
@@ -55,206 +56,291 @@ function Expand-7zipArchive {
-    $ArgList = @('x', "`"$Path`"", "-o`"$DestinationPath`"", '-y')
-    $IsTar = ((strip_ext $Path) -match '\.tar$') -or ($Path -match '\.t[abgpx]z2?$')
-    if (!$IsTar -and $ExtractDir) {
-        $ArgList += "-ir!$ExtractDir\*"
-    }
-    if ($Switches) {
-        $ArgList += (-split $Switches)
-    }
-    switch ($Overwrite) {
-        "All" { $ArgList += "-aoa" }
-        "Skip" { $ArgList += "-aos" }
-        "Rename" { $ArgList += "-aou" }
-    }
-    $Status = Invoke-ExternalCommand $7zPath $ArgList -LogPath $LogPath
-    if (!$Status) {
-        abort "Failed to extract files from $Path.`nLog file:`n  $(friendly_path $LogPath)`n$(new_issue_msg $app $bucket 'decompress error')"
-    }
-    if (!$IsTar -and $ExtractDir) {
-        movedir "$DestinationPath\$ExtractDir" $DestinationPath | Out-Null
-    }
-    if (Test-Path $LogPath) {
-        Remove-Item $LogPath -Force
-    }
-    if ($IsTar) {
-        # Check for tar
-        $Status = Invoke-ExternalCommand $7zPath @('l', "`"$Path`"") -LogPath $LogPath
-        if ($Status) {
-            $TarFile = (Get-Content -Path $LogPath)[-4] -replace '.{53}(.*)', '$1' # get inner tar file name
-            Expand-7zipArchive -Path "$DestinationPath\$TarFile" -DestinationPath $DestinationPath -ExtractDir $ExtractDir -Removal
+    begin {
+        if ((get_config 7ZIPEXTRACT_USE_EXTERNAL)) {
+            try {
+                $7zPath = (Get-Command '7z' -CommandType Application | Select-Object -First 1).Source
+            } catch [System.Management.Automation.CommandNotFoundException] {
+                abort "Cannot find external 7-Zip (7z.exe) while '7ZIPEXTRACT_USE_EXTERNAL' is 'true'!`nRun 'scoop config 7ZIPEXTRACT_USE_EXTERNAL false' or install 7-Zip manually and try again."
+            }
         } else {
-            abort "Failed to list files in $Path.`nNot a 7-Zip supported archive file."
+            $7zPath = Get-HelperPath -Helper 7zip
         }
     }
-    if ($Removal) {
-        # Remove original archive file
-        Remove-Item $Path -Force
+    process {
+        if (!$DestinationPath) {
+            $DestinationPath = Split-Path -Path $Path
+        }
+        for ($i = 0; $i -lt $Path.Length; $i++) {
+            $aPath = $Path[$i]
+            if ($null -eq $DestinationPath[$i]) {
+                $aDestinationPath = $DestinationPath[$i]
+            } else {
+                $aDestinationPath = $DestinationPath[-1]
+            }
+            if ($ExtractDir) {
+                if ($null -eq $ExtractDir[$i]) {
+                    $aExtractDir = $ExtractDir[$i]
+                } else {
+                    $aExtractDir = $ExtractDir[-1]
+                }
+            }
+            $LogPath = "$(Split-Path $aPath)\7zip.log"
+            $ArgList = @('x', "`"$aPath`"", "-o`"$aDestinationPath`"", '-y')
+            $IsTar = ((strip_ext $aPath) -match '\.tar$') -or ($aPath -match '\.t[abgpx]z2?$')
+            if (!$IsTar -and $aExtractDir) {
```

Comment:
`$aExtractDir` is not initialized if `$ExtractDir` is not set.

Reply:
Oop, it should be `-ne` here...